src/grain/harvest.js

```diff
@@ -0,0 +1,220 @@
+ * period, not the current one.
+ */
+import {sum} from "d3-array";
+import {mapToArray} from "../util/map";
+import {type NodeAddressT} from "../core/graph";
+import {type Grain, multiplyFloat, ZERO} from "./grain";
+
+export const HARVEST_VERSION_1 = 1;
+
+export type HarvestStrategy = FastV1 | FairV1;
+
+export type FastV1 = {|
+  +type: "FAST",
+  +version: number,
+  +amount: Grain,
+|};
+
+export type FairV1 = {|
+  +type: "FAIR",
+  +version: number,
+  +amount: Grain,
+|};
+
+export type GrainReceipt = {|+address: NodeAddressT, amount: Grain|};
+
+export type HarvestV1 = {|
+  +type: "HARVEST",
+  +timestampMs: number,
+  +version: number,
+  +strategy: HarvestStrategy,
+  +receipts: $ReadOnlyArray<GrainReceipt>,
+|};
+
+export type CredTimeSlice = {|
+  +intervalEndMs: number,
+  +cred: Map<NodeAddressT, number>,
+|};
+
+export type CredHistory = $ReadOnlyArray<CredTimeSlice>;
+
+/**
+ * Compute a full Harvest given:
+ * - the strategy we're using
+ * - the full cred history for all users
+ * - the lifetime earnings of all users
+ * - the timestamp for the harvest
+ */
+export function harvest(
+  strategy: HarvestStrategy,
+  credHistory: CredHistory,
+  earnings: Map<NodeAddressT, Grain>,
+  timestampMs: number
+): HarvestV1 {
+  const filteredSlices = credHistory.filter(
+    (s) => s.intervalEndMs <= timestampMs
+  );
+  if (!filteredSlices.length) {
+    return {
+      type: "HARVEST",
+      timestampMs,
+      version: HARVEST_VERSION_1,
+      strategy,
+      receipts: [],
+    };
+  }
+
+  const receipts: $ReadOnlyArray<GrainReceipt> = (function () {
+    switch (strategy.type) {
+      case "FAST":
+        if (strategy.version !== 1) {
+          throw new Error(`Unsupported FAST strategy: ${strategy.version}`);
+        }
+        const lastSlice = filteredSlices[filteredSlices.length - 1];
+        return computeFastReceipts(strategy.amount, lastSlice.cred);
+      case "FAIR":
+        if (strategy.version !== 1) {
+          throw new Error(`Unsupported FAIR strategy: ${strategy.version}`);
+        }
+        const totalCred = new Map();
+        for (const {cred} of filteredSlices) {
+          for (const [address, ownCred] of cred.entries()) {
+            const existingCred = totalCred.get(address) || 0;
+            totalCred.set(address, existingCred + ownCred);
+          }
+        }
+        return computeFairReceipts(strategy.amount, totalCred, earnings);
+      default:
+        throw new Error(`Unexpected type ${(strategy.type: empty)}`);
+    }
+  })();
+
+  return {
+    type: "HARVEST",
+    version: HARVEST_VERSION_1,
+    strategy,
+    receipts,
+    timestampMs,
+  };
+}
+
+/**
+ * Split a grain amount in proportion to the provided scores
+ */
+function computeFastReceipts(
+  harvestAmount: Grain,
+  cred: Map<NodeAddressT, number>
+): $ReadOnlyArray<GrainReceipt> {
+  if (harvestAmount < ZERO) {
+    throw new Error(`invalid harvestAmount: ${String(harvestAmount)}`);
+  }
```

Comment:
If we're going to have this sanity check, we should have it for the fair receipts too. No need to add a test IMO, since it's just a sanity check.